gnome-initial-setup/pages/language/cc-language-chooser.c

```diff
@@ -145,10 +147,91 @@ language_widget_new (const char *locale_id,
         return widget->box;
 }
 
+typedef struct {
+        CcLanguageChooser *chooser;
+        GtkWidget         *row;
+} ScrollData;
+
+static inline gint
+get_selected_language_row_y (GtkWidget *row)
+{
+        GtkAllocation alloc;
+
+        gtk_widget_get_allocation (row, &alloc);
+
+        return alloc.y + alloc.height / 2.0;
+}
+
+static gboolean
+update_scroll_position (gpointer user_data)
+{
+        CcLanguageChooser *self;
+        GtkAdjustment *vadjustment;
+        ScrollData *data;
+        GtkWidget *row;
+        gdouble value, real_value;
+        gdouble max, page, page_increment;
+
+        data = user_data;
+        self = data->chooser;
+        row = data->row;
+
+        /* While the scrolled window is not mapped, we keep waiting */
+        if (!gtk_widget_get_mapped (self->priv->language_scrolledwindow))
+                return G_SOURCE_CONTINUE;
+
+        vadjustment = gtk_scrolled_window_get_vadjustment (GTK_SCROLLED_WINDOW (self->priv->language_scrolledwindow));
+        page = gtk_adjustment_get_page_size (vadjustment);
+        max = gtk_adjustment_get_upper (vadjustment);
+
+        page_increment = gtk_adjustment_get_page_increment (vadjustment);
+        real_value = get_selected_language_row_y (row) - page_increment / 2.0;
+        value = gtk_adjustment_get_value (vadjustment);
+
+        gtk_adjustment_set_page_increment (vadjustment, real_value - value);
```

Comment:
The requirement is really only that it appears centered the first time it comes up pre-selected, not that every click centers it, so I don't think the animation is worth the code. I actually think that it is confusing if it centered itself at every click, since that's not what other scrolled lists do.
So I would
- change the code so that it only centers the initial selection and not further clicks
- remove the animation